src/ucp/rma/amo_sw.c

```diff
@@ -194,7 +194,156 @@ DEFINE_AMO_SW_OP(64)
 DEFINE_AMO_SW_FOP(32)
 DEFINE_AMO_SW_FOP(64)
 
-UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_req_handler, (arg, data, length, am_flags),
+typedef struct ucp_atomic_flush_ctx {
+    ucp_atomic_req_hdr_t req_hdr;
+    uint64_t             value;
+    ucp_mem_h            memh;
+    ucp_rkey_h           rkey;
+} ucp_atomic_flush_ctx_t;
+
+static void ucp_amo_sw_loopback_send_reply(ucp_worker_h worker,
+                                           ucp_atomic_flush_ctx_t *flush_ctx)
+{
+    ucp_ep_h reply_ep;
+    ucp_request_t *req;
+    ucs_status_t status;
+
+    reply_ep = ucp_worker_get_ep_by_ptr(worker, flush_ctx->req_hdr.req.ep_ptr);
+
+    if (flush_ctx->req_hdr.req.reqptr == 0) {
+        ucp_rma_sw_send_cmpl(reply_ep);
+    } else {
+        req = ucp_request_get(worker);
+        if (req == NULL) {
+            ucs_error("failed to allocate atomic reply");
+        } else {
+            req->send.ep               = reply_ep;
+            req->send.atomic_reply.req = flush_ctx->req_hdr.req.reqptr;
+            req->send.length           = flush_ctx->req_hdr.length;
+            req->send.uct.func         = ucp_progress_atomic_reply;
+            ucp_request_send(req, 0);
+        }
+    }
+
+    ucp_rkey_destroy(flush_ctx->rkey);
```

Comment:
if the reply didn't finish yet, why is rkey destroyed?

Reply:
AMO is posted and flushed, reply is a common active message with response to initiator